airflow/providers/opsgenie/notifications/opsgenie.py

```diff
@@ -0,0 +1,150 @@
+
+from airflow.exceptions import AirflowOptionalProviderFeatureException
+
+try:
+    from airflow.notifications.basenotifier import BaseNotifier
+except ImportError:
+    raise AirflowOptionalProviderFeatureException(
+        "Failed to import BaseNotifier. This feature is only available in Airflow versions >= 2.6.0"
+    )
+
+from airflow.providers.opsgenie.hooks.opsgenie import OpsgenieAlertHook
+
+if TYPE_CHECKING:
+    from airflow.utils.context import Context
+
+
+class OpsgenieNotifier(BaseNotifier):
+    """
+    This notifier allows you to post alerts to Opsgenie.
+
+    Accepts a connection that has an Opsgenie API key as the connection's password.
+    This notifier sets the domain to conn_id.host, and if not set will default
+    to ``https://api.opsgenie.com``.
+
+    Each Opsgenie API key can be pre-configured to a team integration.
+    You can override these defaults in this notifier.
+
+    .. seealso::
+        For more information on how to use this notifier, take a look at the guide:
+        :ref:`howto/notifier:OpsgenieNotifier`
+
+    :param opsgenie_conn_id: The name of the Opsgenie connection to use
+    :param message: The Message of the Opsgenie alert (templated)
+    :param alias: Client-defined identifier of the alert (templated)
+    :param description: Description field of the alert (templated)
+    :param responders: Teams, users, escalations and schedules that
+        the alert will be routed to send notifications.
+    :param visible_to: Teams and users that the alert will become visible
+        to without sending any notification.
+    :param actions: Custom actions that will be available for the alert.
+    :param tags: Tags of the alert.
+    :param details: Map of key-value pairs to use as custom properties of the alert.
+    :param entity: Entity field of the alert that is
+        generally used to specify which domain alert is related to. (templated)
+    :param source: Source field of the alert. Default value is
+        IP address of the incoming request.
+    :param priority: Priority level of the alert. Default value is P3. (templated)
+    :param user: Display name of the request owner.
+    :param note: Additional note that will be added while creating the alert. (templated)
+    """
+
+    template_fields: Sequence[str] = ("message", "alias", "description", "entity", "priority", "note")
+
+    def __init__(
+        self,
+        *,
+        message: str,
+        opsgenie_conn_id: str = "opsgenie_default",
+        alias: str | None = None,
+        description: str | None = None,
+        responders: list[dict] | None = None,
+        visible_to: list[dict] | None = None,
+        actions: list[str] | None = None,
+        tags: list[str] | None = None,
+        details: dict | None = None,
+        entity: str | None = None,
+        source: str | None = None,
+        priority: str | None = None,
+        user: str | None = None,
+        note: str | None = None,
```

Comment:
I'm not sure maybe from user perspective better to have single `payload: dict | None = None` attribute rather than multiple attributes which later on combined into the single one payload?

Just a question, I'm never use Opsgenie Provider.

Reply:
> Just a question, I'm never use Opsgenie Provider.

For now, me neither, but I'll be a user soon, so that's why I have decided to open this PR.
From a software engineering perspective I feel that this approach works best because it's easier to see what's optional and what isn't, and also the types of each parameter in case you forget. It also makes its usage nearly identical to the `OpsgenieCreateAlertOperator`, so there is no learning curve if someone wants to swap it for this notifier. And besides that there is always dict unpacking if someone already has a payload and wants to use it in this notifier.
I'm open to suggestions, though!

Reply:
Building on what @Taragolis said, just one downside, you always have to keep up with the `Opsgenie` service. In the event that one of the parameters gets deprecated, someone has to make the change here and you would always have to keep up with the changes.

Reply:
Good point, @utkarsharma2. I'll change it to just a payload arg then.

Reply:
If we wont have some hints (optional) we could create [TypedDict](https://docs.python.org/3/library/typing.html#typing.TypedDict), and use it as payload type something like `payload: CreateAlertPayload`

```python
from typing import TypedDict
from typing_extensions import NotRequired  # For compat with Python < 3.11


class CreateAlertPayload(TypedDict):
    message: str
    alias: NotRequired[str]
    description: NotRequired[None]
    responders: NotRequired[list[dict]]
    visible_to: NotRequired[list[dict]]
    actions: NotRequired[list[str]]
    tags: NotRequired[list[str]]
    details: NotRequired[dict]
    entity: NotRequired[str]
    source: NotRequired[str]
    priority: NotRequired[str]
    user: NotRequired[str]
    note: NotRequired[str]
```

---
![image](https://github.com/apache/airflow/assets/3998685/052b828c-3df0-4b2f-8d4d-7af46e1594d4)

---

![image](https://github.com/apache/airflow/assets/3998685/6863482e-1150-4202-a738-a7255b1f073d)

---

![image](https://github.com/apache/airflow/assets/3998685/ad88c56c-a06b-4fcf-89f0-b9e857a12e77)

Reply:
Agreed with the TypedDict, @Taragolis. A couple of questions:
1. Where do you think it could go? I was thinking a common.py file inside the provider so that it could be reused by the `OpsgenieCreateAlertOperator`, but I'm open to suggestions.
2. How do we deal with the `opsgenie_conn_id`?

Reply:
@Taragolis also, how do we make sure that the template_fields will be able to be templated?

Reply:
> Where do you think it could go? I was thinking a common.py file inside the provider so that it could be reused by the OpsgenieCreateAlertOperator, but I'm open to suggestions.

Maybe something opsgenie_typing.py and place CreateAlertPayload there, just suggestion just because: _There are only two hard things in Computer Science: cache invalidation and naming things._

> How do we deal with the opsgenie_conn_id?

Connection id usually (always?) is the separate parameter. so keep it as it is.

Make payload as single dictionary could be better by the reason which @utkarsharma2 mentioned.

> how do we make sure that the template_fields will be able to be templated?

If you put `payload` attribute (or whatever it named) as part of `template_fields` then it would be templated. e.g.

```python
send_opsgenie_notification(
    payload={
        "message": "Something went wrong in {{ dag.dag_id }}!"
        "user": "{{var.value.user }}"
)

Reply:
Won't making payload a template_field also make the iterable fields templated? Would that be ok?

Reply:
Done adding the payload arg.

Reply:
As you can see from the sad path, it's been hard to ensure the schema for payload with TypedDict. Is it the only way to go, @Taragolis?